File: game.cpp
#include <stdio.h>
#include <stdlib.h>

#define WIDTH 7
#define HEIGHT 6

// This struct is what is returned by the isBoardWinnable function. The expectedResult is 1 if
// we can win from this position, 0 if we can tie but not win, and -1 if we're sure to lose.
// The next move is the column we should play to get the best result for us.
typedef struct Tuple {
    int expectedResult;
    int nextMove;
} Tuple;

void printBoard(char** board) {
    for (int i=0; i<HEIGHT; i++) {
        for (int j=0; j<WIDTH; j++)
            printf("%c ", board[i][j]);
        printf("\n");
    }
}

int makeMove(char **board, char player, int column) {
    if (board[0][column] != '-') {
        // illegal move because column is already full
        return -1;
    }

    // the move will change the last empty position in that row
    int rowToChange;
    for (rowToChange=0; rowToChange<HEIGHT-1 && board[rowToChange+1][column]=='-'; rowToChange++) {}

    board[rowToChange][column] = player;

    return 1;
}

char checkWinnerSmall(char** board) {
    // vertically
    for (int i=0; i<HEIGHT-2; i++)
        for (int j=0; j<WIDTH; j++) {
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j]
                && board[i][j] == board[i+2][j]
            ) return board[i][j];
        }

    // horizontally
    for (int i=0; i<HEIGHT; i++)
        for (int j=0; j<WIDTH-2; j++) {
            if (
                board[i][j] != '-'
                && board[i][j] == board[i][j+1]
                && board[i][j] == board[i][j+2]
            ) return board[i][j];
        }

    // diagonally (primary)
    for (int i=0; i<HEIGHT-2; i++)
        for (int j=0; j<WIDTH-2; j++)
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j+1]
                && board[i][j] == board[i+2][j+2]
            ) return board[i][j];

    // diagonally (secondary)
    for (int i=0; i<HEIGHT-2; i++)
        for (int j=2; j<WIDTH; j++)
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j-1]
                && board[i][j] == board[i+2][j-2]
            ) return board[i][j];

    // If no one has won and all the columns are full we return 't', for tie,
    // otherwise we return '-', meaning there are still valid movements to be made.
    for (int j=0; j<WIDTH; j++)
        if (board[0][j] == '-') return '-';

    return 't';
}

char checkWinner(char** board) {
    // vertically
    for (int i=0; i<HEIGHT-3; i++)
        for (int j=0; j<WIDTH; j++) {
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j]
                && board[i][j] == board[i+2][j]
                && board[i][j] == board[i+3][j]
            ) return board[i][j];
        }

    // horizontally
    for (int i=0; i<HEIGHT; i++)
        for (int j=0; j<WIDTH-3; j++) {
            if (
                board[i][j] != '-'
                && board[i][j] == board[i][j+1]
                && board[i][j] == board[i][j+2]
                && board[i][j] == board[i][j+3]
            ) return board[i][j];
        }

    // diagonally (primary)
    for (int i=0; i<HEIGHT-3; i++)
        for (int j=0; j<WIDTH-3; j++)
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j+1]
                && board[i][j] == board[i+2][j+2]
                && board[i][j] == board[i+3][j+3]
            ) return board[i][j];

    // diagonally (secondary)
    for (int i=0; i<HEIGHT-3; i++)
        for (int j=3; j<WIDTH; j++)
            if (
                board[i][j] != '-'
                && board[i][j] == board[i+1][j-1]
                && board[i][j] == board[i+2][j-2]
                && board[i][j] == board[i+3][j-3]
            ) return board[i][j];

    // If no one has won and all the columns are full we return 't', for tie,
    // otherwise we return '-', meaning there are still valid movements to be made.
    for (int j=0; j<WIDTH; j++)
        if (board[0][j] == '-') return '-';

    return 't';
}

Tuple isBoardWinnable(char** board, char player, char enemy, int origin) {
    static int count = 0;
    //printf("\n\n%d\n", count);
    int current = count;
    count = count + 1;

    //printf("base board\n");
    //printBoard(board);

    // This function receives a board state, and recursively simulates movements to check if
    // there's a sure way to win from the received board. If there is, this returns 1, otherwise
    // returns -1 if we're sure to lose and 0 if we can tie.

    Tuple returnTuple;
    returnTuple.nextMove = -1;

    // The first thing we check is whether we've already won, lost or tied:
    char winner = checkWinner(board);
    if (winner == player) {
        returnTuple.expectedResult = 1;
        //printf("%d returning win to %d\n", current, origin);
        return returnTuple;
    }
    if (winner == 't') {
        returnTuple.expectedResult = 0;
        //printf("%d returning tie to %d\n", current, origin);
        return returnTuple;
    }
    if (winner != '-') {
        returnTuple.expectedResult = -1;
        //printf("%d returning loss to %d\n", current, origin);
        return returnTuple;
    }

    // We'll create one board to simulate each possible movement
    // Since there's one possible movement per column, we create WIDTH simulated boards
    char*** simulatedBoards;
    simulatedBoards = (char***)malloc(WIDTH*sizeof(char**));
    for (int k=0; k<WIDTH; k++) {
        simulatedBoards[k] = (char**)malloc(HEIGHT*sizeof(char*));
        for (int i=0; i<HEIGHT; i++)
            simulatedBoards[k][i] = (char*)malloc(WIDTH*sizeof(char));
    }

    // And we'll initialize them with the current board state
    for (int k=0; k<WIDTH; k++)
        for (int i=0; i<HEIGHT; i++)
            for (int j=0; j<WIDTH; j++)
                simulatedBoards[k][i][j] = board[i][j];

    // This is an array in the format [1, 1, -1, 1, ...], where a 1 in a given index means that
    // a movement in that column is a valid one. If the column is full, we'll have a -1 in that column's
    // index.
    int validMovements[WIDTH];
    for (int k=0; k<WIDTH; k++)
        validMovements[k] = makeMove(simulatedBoards[k], player, k);

    for (int k=0; k<WIDTH; k++) {
        if (validMovements[k] > 0) {
            //printf("simulated board %d\n", k);
            //printBoard(simulatedBoards[k]);
        }
    }

    // If we win in any of the simulated boards, the answer is yes
    for (int k=0; k<WIDTH; k++)
        // We check if the movement is valid to prevent unnecessary calculations
        if (validMovements[k]>0 && checkWinner(simulatedBoards[k]) == player) {
            returnTuple.expectedResult = 1;
            returnTuple.nextMove = k;
            //printf("\n%d returning win %d %d to %d\n", current, returnTuple.expectedResult, returnTuple.nextMove, origin);
            return returnTuple;
        }

    // Otherwise we need to simulate all possible responses to each movement we can make,
    // and choose the one that gives us the best result, assuming the other player will always
    // choose the best move for them.

    int bestPossibleResult = -2; // We start assuming we'll lose until we find a better possibility
    for (int k=0; k<WIDTH; k++) {
        char*** simulatedResponses;
        simulatedResponses = (char***)malloc(WIDTH*sizeof(char**));
        for (int r=0; r<WIDTH; r++) {
            simulatedResponses[r] = (char**)malloc(HEIGHT*sizeof(char*));
            for (int i=0; i<HEIGHT; i++)
                simulatedResponses[r][i] = (char*)malloc(WIDTH*sizeof(char));
        }

        // Simulating the possible responses is similar to simulating the possible movements.
        for (int r=0; r<WIDTH; r++)
            for (int i=0; i<HEIGHT; i++)
                for (int j=0; j<WIDTH; j++)
                    simulatedResponses[r][i][j] = simulatedBoards[k][i][j];

        int validResponses[WIDTH];
        for (int r=0; r<WIDTH; r++) {
            if (validMovements[k] < 0) {
                validResponses[r] = -1; // If the original movement was invalid so is any response
            } else {
                validResponses[r] = makeMove(simulatedResponses[r], enemy, r);
            }
        }

        for (int r=0; r<WIDTH; r++) {
            if (validResponses[r] > 0) {
                //printf("simulated response %d (current %d)\n", r, current);
                //printBoard(simulatedResponses[r]);
            }
        }

        // The enemy will choose what's worst for us, so here we need get the minimum simulated value
        int worstPossibleResult = 2; // We need to start with an invalid value that is bigger than all the options
        Tuple isWinnable;
        for (int r=0; r<WIDTH; r++)
            if (validResponses[r] > 0) {
                //printf("\n calling for response %d", r);
                isWinnable = isBoardWinnable(simulatedResponses[r], player, enemy, current);
                if (isWinnable.expectedResult < worstPossibleResult) worstPossibleResult = isWinnable.expectedResult;
            }

        // If the worst possible result for this move is 1, we don't need to keep on searching,
        // because we've already found a move that is sure to make us win
        if (worstPossibleResult == 1) {
            returnTuple.expectedResult = 1;
            returnTuple.nextMove = k;
            //printf("\n%d returning rec1 %d %d to %d\n", current, returnTuple.expectedResult, returnTuple.nextMove, origin);
            return returnTuple;
        }

        // worstPossibleResult still being 2 means all moves were invalid
        if (worstPossibleResult < 2 && worstPossibleResult > bestPossibleResult) {
            bestPossibleResult = worstPossibleResult;
            returnTuple.expectedResult = bestPossibleResult;
            returnTuple.nextMove = k;
        }
    }

    // If my move was the last possible one and I didn't win, return a tie with that movement
    if (returnTuple.nextMove == -1) {
        int lastMove;
        for (lastMove = 0; lastMove<WIDTH+1 && validMovements[lastMove]<0; lastMove++) {}
        returnTuple.expectedResult = 0;
        returnTuple.nextMove = lastMove;
    }
    //printf("\n%d returning rec2 %d %d to %d\n", current, returnTuple.expectedResult, returnTuple.nextMove, origin);
    return returnTuple;
}

int enemyMove(char** board, char player, char enemy) {
    return isBoardWinnable(board, player, enemy, -1).nextMove;
}

int main() {
    char** board;
    const char players[2] = {'x', 'o'};

    // initialize empty board
    board = (char**)malloc(HEIGHT*sizeof(char*));
    for (int i=0; i<HEIGHT; i++)
        board[i] = (char*)malloc(WIDTH*sizeof(char));

    for (int i=0; i<HEIGHT; i++)
        for (int j=0; j<WIDTH; j++)
            board[i][j] = '-';

    char currentPlayerIndex = 1;
    int chosenColumn;
    char winner = '-';
    while (winner == '-') {
        char currentPlayer = players[currentPlayerIndex];
        char enemy = players[(currentPlayerIndex+1)%2];

        if (currentPlayerIndex == 1) {
            // non-human player's turn
            makeMove(board, currentPlayer, enemyMove(board, currentPlayer, enemy));
        } else {
            printf("\nTabuleiro:\n");
            printBoard(board);

            printf("\nSua vez!");
            printf("\nDigite a coluna em que voce quer jogar: ");
            scanf(" %d", &chosenColumn);
            while (chosenColumn < 0 || chosenColumn >= WIDTH) {
                printf("Coluna invalida. Digite um numero entre 0 e %d: ", HEIGHT-1);
                scanf(" %d", &chosenColumn);
            }

            int isMoveValid = makeMove(board, currentPlayer, chosenColumn);
            while (isMoveValid < 0) {
                printf("Coluna cheia. Escolha outra: ");
                scanf(" %d", &chosenColumn);
                while (chosenColumn < 0 || chosenColumn >= WIDTH) {
                    printf("Coluna invalida. Digite um numero entre 0 e %d: ", HEIGHT-1);
                    scanf(" %d", &chosenColumn);
                }

                isMoveValid = makeMove(board, currentPlayer, chosenColumn);
            }
        }

        currentPlayerIndex = (currentPlayerIndex + 1) % 2;
        winner = checkWinner(board);
    }

    printBoard(board);
    printf("O vencedor eh o jogador %c!", winner);
}

File: game.py
import pprint
import math

width = 4
height = 4
players = ['x', 'o']

def print_board(board):
    for row in range(height):
        print(" ".join(board[row]))


def make_move(board, player_idx, column):
    player = players[player_idx]

    if board[0][column] != '-':
        # column full
        return False

    last_empty_row = 0
    while last_empty_row < height - 1 and board[last_empty_row+1][column] == '-':
        last_empty_row = last_empty_row + 1

    board[last_empty_row][column] = player
    return True


def check_winner_small(board):
    # horizontally
    for row in range(height):
        for col in range(width-2):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row][col+1]
                and board[row][col] == board[row][col+2]
            ):
                return board[row][col]

    # vertically
    for row in range(height-2):
        for col in range(width):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col]
                and board[row][col] == board[row+2][col]
            ):
                return board[row][col]

    # diagonally (primary)
    for row in range(height-2):
        for col in range(width-2):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col+1]
                and board[row][col] == board[row+2][col+2]
            ):
                return board[row][col]

    # diagonally (secondary)
    for row in range(height-2):
        for col in range(2, width):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col-1]
                and board[row][col] == board[row+2][col-2]
            ):
                return board[row][col]

    # if there are empty space no one won yet
    for col in range(width):
        if board[0][col] == '-':
            return '-'

    # if there aren't the game is over on a tie
    return 't'


def check_winner(board):
    # horizontally
    for row in range(height):
        for col in range(width-3):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row][col+1]
                and board[row][col] == board[row][col+2]
                and board[row][col] == board[row][col+3]
            ):
                return board[row][col]

    # vertically
    for row in range(height-3):
        for col in range(width):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col]
                and board[row][col] == board[row+2][col]
                and board[row][col] == board[row+3][col]
            ):
                return board[row][col]

    # diagonally (primary)
    for row in range(height-3):
        for col in range(width-3):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col+1]
                and board[row][col] == board[row+2][col+2]
                and board[row][col] == board[row+3][col+3]
            ):
                return board[row][col]

    # diagonally (secondary)
    for row in range(height-3):
        for col in range(3, width):
            if (
                board[row][col] != '-'
                and board[row][col] == board[row+1][col-1]
                and board[row][col] == board[row+2][col-2]
                and board[row][col] == board[row+3][col-3]
            ):
                return board[row][col]

    # if there are empty space no one won yet
    for col in range(width):
        if board[0][col] == '-':
            return '-'

    # if there aren't the game is over on a tie
    return 't'

def get_possible_next_positions(board, player_idx):
    simulated_boards = []
    for col in range(width):
        if board[0][col] == '-':
            new_board = [[board[i][j] for j in range(width)] for i in range(height)]
            make_move(new_board, player_idx, col)
            simulated_boards.append((new_board, col))

    return simulated_boards

def minmax(board, player_idx, alpha, beta):
    winner = check_winner_small(board) 
    if winner != '-':
        return {
            "eval": {
                players[0]: 1,
                players[1]: -1,
                't': 0,
            }[winner],
            "move": -1
        }

    possible_boards = get_possible_next_positions(board, player_idx)

    if player_idx == 0:
        max_eval = -math.inf
        next_move = -1
        for next_board in possible_boards:
            if alpha < beta:
                simulation = minmax(next_board[0], (player_idx + 1) % 2, alpha, beta)
                if simulation["eval"] > max_eval:
                    max_eval = simulation["eval"]
                    next_move = next_board[1]
                alpha = max(alpha, simulation["eval"])
        return {
            "eval": max_eval,
            "move": next_move
        }
        
    if player_idx == 1:
        min_eval = math.inf
        next_move = -1
        for next_board in possible_boards:
            if alpha < beta:
                simulation = minmax(next_board[0], (player_idx + 1) % 2, alpha, beta)
                if simulation["eval"] < min_eval:
                    min_eval = simulation["eval"]
                    next_move = next_board[1]
                beta = min(beta, simulation["eval"])
        return {
            "eval": min_eval,
            "move": next_move
        }


def game():
    board = [['-' for _ in range(width)] for _ in range(height)]
    player_idx = 1
    while check_winner_small(board) == '-':
        if player_idx == 1:
            # computer turn
            simulation = minmax(board, player_idx, -math.inf, math.inf)
            make_move(board, player_idx, simulation["move"])

        else:
            print("Tabuleiro:")
            print_board(board)

            print("\nSua vez!", players[player_idx])
            print("Digita a coluna em que quer jogar: ")
            column = int(input())

            while column < 0 or column >= width:
                print("Coluna inválida. Digite um número entre 0 e {}: ".format(width-1))
                column = int(input())

            valid_move = make_move(board, player_idx, column)
            while not valid_move:
                print("Coluna cheia. Escolha outra: ")
                column = int(input())
                while column < 0 or column >= width:
                    print("Coluna inválida. Digite um número entre 0 e {}: ".format(width-1))
                    column = int(input())

                valid_move = make_move(board, player_idx, column)

        player_idx = (player_idx + 1) % 2

    print_board(board)
    print("Vencedor: ", check_winner_small(board))

game()